Add Sensei Pro promo backend for Sensei Home.

aaronfc opened the pull request:
Fixes #5744 

### Changes proposed in this Pull Request
* Changed interface for promotional banner so that it is a little bit more generic and avoid any references to sensei pro.
* Logic to show/hide banner is based on a new filter `sensei_home_promo_banner_show`.

### Testing instructions
- Run curl similar to: `curl -sX GET --user aaron:development --insecure 'https://test2.local/wp-json/sensei-internal/v1/home' | jq .promo_banner`. ([Check here to make curl authentication easier in local](https://evergreenp2.wordpress.com/2022/03/03/testing-cheatsheet/#test-rest-api-in-local-easily))
- Check that output is as expected (visible by default).
- Disable `sensei_home_promo_banner_show` filter with a snippet.
- Run curl again.
- Check that output is correct (not visible).

### Screenshot / Video
**By default**
```
$ curl -sX GET --user aaron:development --insecure 'https://test2.local/wp-json/sensei-internal/v1/home' | jq .promo_banner
{
  "is_visible": true
}
```
**With filter disabled**
```
$ curl -sX GET --user aaron:development --insecure 'https://test2.local/wp-json/sensei-internal/v1/home' | jq .promo_banner
{
  "is_visible": false
}
```

aaronfc committed:
Add Sensei Pro promo backend for Sensei Home.

fjorgemota reviewed (commented):
Given the dependency of this code on the `Sensei_Pro_Detector`, this code also has the same issue that I shared here: https://github.com/Automattic/sensei/pull/5786#pullrequestreview-1122834605 

Here, however, I feel like we can change completely the approach: Instead of having a filter that says "well, Sensei Pro is installed! Let's hide the ad!", we can have Sensei Pro change the filter the output of that REST API directly and change that flag directly. 

Given that we will need to add some additional licensing data on that endpoint anyway, this looks like a good way to also hide the promo if Sensei Pro is properly installed.

Does that make sense?

aaronfc commented:
> Here, however, I feel like we can change completely the approach: Instead of having a filter that says "well, Sensei Pro is installed! Let's hide the ad!", we can have Sensei Pro change the filter the output of that REST API directly and change that flag directly.

🤔  I don't like very much the idea of Sensei Pro overriding directly the REST API with some specific flag. I would rather continue with the Sensei Pro detection proposal. See my comment and changes in #5742 

Otherwise we would be coupling Sensei Pro to the specific details on the REST API, which doesn't sound very reasonable to me.

fjorgemota commented on includes/admin/home/promo/class-sensei-home-sensei-pro-promo-provider.php:
```diff
@@ -0,0 +1,40 @@
+/**
+ * Class responsible for returning if the promo for Sensei Pro must be shown or not in the Sensei Home screen.
+ */
+class Sensei_Home_Sensei_Pro_Promo_Provider {
+
+	/**
+	 * Sensei Pro detector.
+	 *
+	 * @var Sensei_Pro_Detector
+	 */
+	private $pro_detector;
+
+	/**
+	 * Class constructor.
+	 *
+	 * @param Sensei_Pro_Detector $pro_detector The Sensei Pro detector.
+	 */
+	public function __construct( Sensei_Pro_Detector $pro_detector ) {
+		$this->pro_detector = $pro_detector;
+	}
+
+
+	/**
+	 * Returns a single boolean representing whether the promo must be shown or not.
+	 *
+	 * @return boolean
+	 */
+	public function get(): bool {
+		$is_pro_loaded = $this->pro_detector->is_loaded();
+		return ! $is_pro_loaded;
+	}
+}
```
Do we need this abstraction just to have a single `! $this->pro_detector->is_loaded();`?

aaronfc replied:
This has been refactored. Let me know if it's still that concerning to you :)

fjorgemota reviewed (approved):
Hey there. The code looks generally OK and it works well. But I left a comment about the abstraction on the `Sensei_Pro_Promo_Provider` which, IMHO, isn't needed.

Personally, I'm worried about having such an abstraction considering that, at least in the current requirements, the **only** thing we worry about is hiding the ad when Sensei Pro is activated, so I'm not sure if we'll have other requirements here just to have a single negation here...

Sure, we have the tests for it (and I really like having automated tests!), but again, right now, it's just a single check on whether Sensei Pro is activated or not, so I'm not sure if there's value in having all this extra abstraction...

However, given the code works, I'm already approving it anyway. :)

aaronfc committed:
Update with `add/home-help` after `Sensei_Pro_Detector` removal.

aaronfc committed:
Refactor Sensei Pro banner to generic promotional banner.

- Remove direct references to Sensei Pro.
- Changed interface for Sensei Home REST API to include a `promo_banner`
  section with an `is_visible` boolean.
- Logic now relies on new `sensei_home_promo_banner_show` filter.

aaronfc committed:
Update with `add/home-help`

aaronfc committed:
Fix access level for tearDown.

aaronfc committed:
Add unit test for promo banner class.

aaronfc commented:
> Hey there. The code looks generally OK and it works well. But I left a comment about the abstraction on the Sensei_Pro_Promo_Provider which, IMHO, isn't needed.

Hey @fjorgemota thanks for your feedback. After reimplementing the Help section to use a filter I have done the same for the promotional banner. Additionally I have removed references to Sensei Pro and made it slightly more generic.
I have still kept the Provider abstraction and added a `Sensei_Home_Promo_Banner` which now represents the full information for the banner (which currently is just an `is_visible`).

I understand your point and we could probably make it work checking for the filter directly in the controller but I feel like having this "framework" is a better bet for maintainability and future developments. Let me know if you have still such concerns about this layer after the reimplementation :)

jom commented:
@aaronfc Sorry for the late review. I get your point about having a framework and your tests show the clean separation of concerns, but I also think it is worth making sure we're not adding too much complexity--especially given that we're going to have to figure out how to take a good chunk of this and copy it in Sensei Blocks (not this PR's work, obviously) to load when Sensei isn't installed. 

With the value->provider->mapper architecture here, are you thinking that these providers will be used anywhere else besides this endpoint? I agree that it'd be nice to separate the REST API controller concern from the concern of collecting the data for the endpoint response, but _if_ we don't foresee these providers being used elsewhere could they be aware of the required data structure of what they're providing?

jom commented on includes/class-sensei-autoloader.php:
```diff
@@ -82,6 +82,7 @@ public function __construct() {
 			new Sensei_Autoloader_Bundle( 'reports/overview/services' ),
 			new Sensei_Autoloader_Bundle( 'admin/home/quick-links' ),
 			new Sensei_Autoloader_Bundle( 'admin/home/help' ),
+			new Sensei_Autoloader_Bundle( 'admin/home/promo-banner' ),
```
I started this for now in my PR #5820 as well, but we might want to move away from these bundlers. I think it slows down the autoloader because every time the autoloader is used, it has to go through each of these and do potentially **two** `file_exists` checks for any unloaded class in the WordPress instance.

aaronfc replied:
Yeah.. At first I started to manually add the entries one by one but later used this for simplicity. You are right about performance... Food for thought: It would be nice if we could autogenerate the list of files instead of manually adding them.

Just to confirm I understand, you are suggesting removing this bundle and adding all the files one by one below in the list of files. Right?

aaronfc commented:
Hey @jom thanks for the review!

> @aaronfc Sorry for the late review. I get your point about having a framework and your tests show the clean separation of concerns,

Yeah, the main goals for doing all of this was, as you pointed out, to have a clear separation of responsibilities which tends to make testing easier/cleaner.

> but I also think it is worth making sure we're not adding too much complexity--especially given that we're going to have to figure out how to take a good chunk of this and copy it in Sensei Blocks (not this PR's work, obviously) to load when Sensei isn't installed.

I believe you have already raised your concerns about this before but I am struggling to understand what's the issue you foresee. I haven't thought much about it but I think that Sensei Home will be the same but with some "parts" missing. I would remove those pieces from the REST API response either in the Controller itself or at Provider level (maybe we can somehow "add" providers intro the controller using a filter? So we can add only relevant providers to the controller 🤔 ).
Is there any specific issue I am missing and you would like to point out?

> With the value->provider->mapper architecture here, are you thinking that these providers will be used anywhere else besides this endpoint? I agree that it'd be nice to separate the REST API controller concern from the concern of collecting the data for the endpoint response, but if we don't foresee these providers being used elsewhere could they be aware of the required data structure of what they're providing?

This is the kind of feedback I was expecting 😄  yes, I agree that this might be "too much" and I had my own doubts if it was worth the effort. I understand you are suggesting to:
- Remove the mapper layer and include that "mapping" logic into the Providers themselves.
- If providers are returning what the controller expects, then there might not be any need for a "domain" layer (I can remove data classes).

Is the above correct? @jom And, would you be open for me to integrate all of this and refactor it (simplify as you suggested) in a separate branch?

fjorgemota reviewed (approved):
Hey @aaronfc, thanks for your reply to @jom. 

I agree with @jom here that this might be too much, and that's what I'm mostly worried about. 

I sent you a message on Slack to discuss my views regarding the simplicity that we're trying to bring to this PR.

Considering that you proposed to work more on simplifying this code on another PR and that the current PR works, I'm already approving it. But I'd appreciate it a lot if you send another PR working on the points I brought on Slack. :) 

Thanks!

jom commented:
@aaronfc Yup, I'm happy for you to merge and then refactor in a separate PR. I really appreciate your work so far on getting this endpoint data all prepped.

> Is there any specific issue I am missing and you would like to point out?

re: Sensei Blocks:
- Physical file sync: Shared files will need to physically synced from this repo to Sensei Blocks. It'd be nice if we were able to keep the shared files and code minimal and stable so we don't have to make many changes here. 
- Loading preference: We'll need to find a safe way to load from Sensei if it is active and fall back to Sensei Blocks if not. Alternatively, we could possibly auto-namespace when syncing the files. If not, we'll probably need protections in place before we load files to not introduce a conflict.

I think Sensei Blocks will have its own REST API controller, just because its dependencies will be different, but will need to provide a similar schema so the shared frontend will be able to correctly handle the data.

aaronfc commented:
I am merging and will work on simplifying in a separate branch as we agreed.

I will follow up with you, @jom, on slack about the Sensei Blocks approach.
